File: harness/oracle/cases/text-detection.ts
/**
 * Tier-1 oracle case — the file text/MIME sniffing leaf (v4
 * `lib/files/text-detection.ts`).
 *
 * Drives v4's REAL `isTextContent` / `getMimeTypeFromExtension` /
 * `isTextMimeType` / `detectTextContent` / `getBestMimeType` over a fixed corpus
 * and emits one NDJSON row per case, each tagged with a `kind` discriminator the
 * Rust port (`quilltap_core::files::text_detection`) dispatches on. Every row is
 * a tier-1 exact-equality check.
 *
 * Buffers are carried as `bytes: number[]` and reconstructed via
 * `Buffer.from(bytes)`. The corpus covers: content sniffing (empty, plain ASCII,
 * 2 nulls ok, 3 nulls binary, exactly-10%-non-printable, high-byte UTF-8, a
 * >8192 buffer whose text head is sampled while its binary tail is ignored);
 * extension inference (compound `.d.ts`/`.min.js`/`.env.local`, uppercase `.PY`,
 * special files `Dockerfile`/`Makefile`/`README`, a path basename, a bare
 * `noext`, an unknown `.zzz`); MIME predicates; full detection (generic +
 * extension, non-generic, binary + generic, text + generic + no extension); and
 * best-MIME fallback (present / null / empty-string).
 *
 * Run from the v4 server checkout:
 *   cd ~/source/quilltap-server
 *   npx tsx ~/source/quilltap-v5/harness/oracle/cases/text-detection.ts \
 *     > /tmp/oracle-text-detection.ndjson
 */

import {
  detectTextContent,
  getBestMimeType,
  getMimeTypeFromExtension,
  isTextContent,
  isTextMimeType,
  type TextDetectionResult,
} from '@/lib/files/text-detection';

const emit = (row: unknown) => process.stdout.write(JSON.stringify(row) + '\n');

// --- isTextContent -----------------------------------------------------------
const asciiBytes = Array.from('hello world\n', (c) => c.charCodeAt(0));
// 2 nulls interleaved with text — the 3rd-null short-circuit is NOT hit; keep
// the buffer long enough that 2/len stays below 0.10 so it resolves as text.
const twoNulls = Array.from('a'.repeat(30), (c) => c.charCodeAt(0));
twoNulls[2] = 0;
twoNulls[17] = 0;
const threeNulls = [0, 0, 0, ...Array.from('a'.repeat(30), (c) => c.charCodeAt(0))];
// Exactly 10% non-printable: 1 control byte out of 10 → ratio 0.10, NOT < 0.10.
const tenPercent = [...Array.from('a'.repeat(9), (c) => c.charCodeAt(0)), 0x01];
const highBytes = [0xff, 0xfe, 0x80, 'a'.charCodeAt(0)];
// >8192: a text head (sampled) followed by a binary tail (ignored).
const bigTextHeadBinaryTail = [
  ...Array.from({ length: 8192 }, () => 'a'.charCodeAt(0)),
  0x00,
  0x00,
  0x00,
  0x00,
];

const contentCases: Array<{ id: string; bytes: number[] }> = [
  { id: 'empty', bytes: [] },
  { id: 'plain-ascii', bytes: asciiBytes },
  { id: 'two-nulls-ok', bytes: twoNulls },
  { id: 'three-nulls-binary', bytes: threeNulls },
  { id: 'exactly-ten-percent', bytes: tenPercent },
  { id: 'high-bytes-utf8', bytes: highBytes },
  { id: 'big-text-head-binary-tail', bytes: bigTextHeadBinaryTail },
];
for (const { id, bytes } of contentCases) {
  emit({ kind: 'isTextContent', id, bytes, out: isTextContent(Buffer.from(bytes)) });
}

// --- getMimeTypeFromExtension ------------------------------------------------
const mimeCases = [
  'foo.ts',
  'component.d.ts',
  'bundle.min.js',
  'App.PY',
  'Dockerfile',
  'Makefile',
  'README',
  'app.env.local',
  'noext',
  'weird.zzz',
  'dir/sub/file.py',
  'archive.tar.gz',
];
for (const filename of mimeCases) {
  emit({
    kind: 'getMime',
    id: `mime:${filename}`,
    filename,
    out: getMimeTypeFromExtension(filename),
  });
}

// --- isTextMimeType ----------------------------------------------------------
const mimeTypeCases = [
  'text/plain',
  'text/x-python',
  'application/json',
  'application/xml',
  'application/javascript',
  'application/typescript',
  'application/x-javascript',
  'application/x-typescript',
  'application/x-sh',
  'application/x-shellscript',
  'application/graphql',
  'application/octet-stream',
  'image/png',
];
for (const mime of mimeTypeCases) {
  emit({ kind: 'isTextMime', id: `textmime:${mime}`, mime, out: isTextMimeType(mime) });
}

// --- detectTextContent -------------------------------------------------------
const detectCases: Array<{
  id: string;
  bytes: number[];
  filename: string;
  provided: string;
}> = [
  // Generic octet-stream + .py extension → detect text/x-python.
  { id: 'generic-with-ext', bytes: asciiBytes, filename: 'script.py', provided: 'application/octet-stream' },
  // Non-generic provided type → detected stays null; still plain text.
  { id: 'nongeneric-provided', bytes: asciiBytes, filename: 'notes.txt', provided: 'text/markdown' },
  // Binary content + generic → detected null, not plain text.
  { id: 'binary-generic', bytes: threeNulls, filename: 'blob.bin', provided: 'application/octet-stream' },
  // Text content + generic + no known extension → text/plain.
  { id: 'text-generic-noext', bytes: asciiBytes, filename: 'mystery', provided: '' },
  // Text content + generic + unknown extension → text/plain (no extension mime).
  { id: 'text-generic-unknown-ext', bytes: asciiBytes, filename: 'data.zzz', provided: 'application/x-unknown' },
  // Non-generic image type over binary bytes → not plain text, detected null.
  { id: 'image-binary', bytes: threeNulls, filename: 'pic.png', provided: 'image/png' },
];
for (const { id, bytes, filename, provided } of detectCases) {
  const out = detectTextContent(Buffer.from(bytes), filename, provided);
  emit({ kind: 'detect', id, bytes, filename, provided, out });
}

// --- getBestMimeType ---------------------------------------------------------
const bestCases: Array<{ id: string; detectedMimeType: string | null; provided: string }> = [
  { id: 'detected-present', detectedMimeType: 'text/x-python', provided: 'application/octet-stream' },
  { id: 'detected-null', detectedMimeType: null, provided: 'application/octet-stream' },
  { id: 'detected-empty', detectedMimeType: '', provided: 'application/octet-stream' },
];
for (const { id, detectedMimeType, provided } of bestCases) {
  const result: TextDetectionResult = {
    isPlainText: false,
    detectedMimeType,
    encoding: 'utf-8',
  };
  emit({ kind: 'getBest', id, detectedMimeType, provided, out: getBestMimeType(result, provided) });
}

File: crates/quilltap-core/src/files/text_detection.rs
//! Text content / MIME sniffing — a byte-exact port of v4
//! `lib/files/text-detection.ts`.
//!
//! Used during file upload to (1) detect whether a file is readable as text,
//! (2) infer a better MIME type when the browser hands us a generic one, and
//! (3) support previewing and LLM access to code/text files. This is the
//! upload-half leaf; it ships with its own tier-1 differential
//! (`text_detection_equivalence`).
//!
//! Fidelity notes carried from v4:
//! * [`is_text_content`] samples at most [`SAMPLE_SIZE`] bytes; an empty file IS
//!   text; the 3rd null byte short-circuits to binary; the non-printable ratio
//!   is a STRICT `<` against [`NON_PRINTABLE_THRESHOLD`], denominator =
//!   `sample_length`. A null byte is itself non-printable, so it increments both
//!   counters.
//! * [`is_printable_or_whitespace`] treats every high byte (`>= 0x80`) as
//!   printable with NO UTF-8 validation (v4's "could be a valid continuation").
//! * [`get_mime_type_from_extension`]'s compound pass walks
//!   `EXTENSION_MIME_MAP` in INSERTION ORDER and returns the first `ends_with`
//!   hit — so `component.d.ts` resolves via `.ts` (the first entry). The slice
//!   ordering is load-bearing; do not turn it into a map.

use regex::Regex;
use std::sync::LazyLock;

/// Bytes to sample for text detection (v4 `SAMPLE_SIZE`).
pub const SAMPLE_SIZE: usize = 8192;

/// Maximum ratio of non-printable characters to still count as text
/// (v4 `NON_PRINTABLE_THRESHOLD`).
pub const NON_PRINTABLE_THRESHOLD: f64 = 0.10;

/// Result of text-detection analysis (v4 `TextDetectionResult`).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TextDetectionResult {
    /// Whether the file content appears to be plain text.
    pub is_plain_text: bool,
    /// Inferred MIME type based on extension/content (if better than provided).
    pub detected_mime_type: Option<String>,
    /// Detected or assumed encoding (always `"utf-8"`).
    pub encoding: String,
}

/// File extension → MIME type mapping (v4 `EXTENSION_MIME_MAP`), a slice so the
/// INSERTION ORDER of the compound `ends_with` pass is preserved verbatim.
///
/// (`.R` becomes unreachable after the caller lowercases, but it is kept for
/// fidelity with the v4 object.)
const EXTENSION_MIME_MAP: &[(&str, &str)] = &[
    // TypeScript/JavaScript
    (".ts", "text/typescript"),
    (".tsx", "text/typescript"),
    (".js", "text/javascript"),
    (".jsx", "text/javascript"),
    (".mjs", "text/javascript"),
    (".cjs", "text/javascript"),
    // Web
    (".html", "text/html"),
    (".htm", "text/html"),
    (".css", "text/css"),
    (".scss", "text/x-scss"),
    (".sass", "text/x-sass"),
    (".less", "text/x-less"),
    (".vue", "text/x-vue"),
    (".svelte", "text/x-svelte"),
    // Data formats
    (".json", "application/json"),
    (".xml", "application/xml"),
    (".yaml", "text/yaml"),
    (".yml", "text/yaml"),
    (".toml", "text/toml"),
    (".csv", "text/csv"),
    (".tsv", "text/tab-separated-values"),
    // Documentation
    (".md", "text/markdown"),
    (".markdown", "text/markdown"),
    (".txt", "text/plain"),
    (".text", "text/plain"),
    (".log", "text/plain"),
    (".rst", "text/x-rst"),
    // Programming languages
    (".py", "text/x-python"),
    (".pyw", "text/x-python"),
    (".rb", "text/x-ruby"),
    (".rs", "text/x-rust"),
    (".go", "text/x-go"),
    (".java", "text/x-java"),
    (".kt", "text/x-kotlin"),
    (".kts", "text/x-kotlin"),
    (".scala", "text/x-scala"),
    (".swift", "text/x-swift"),
    (".c", "text/x-c"),
    (".h", "text/x-c"),
    (".cpp", "text/x-c++"),
    (".hpp", "text/x-c++"),
    (".cc", "text/x-c++"),
    (".cxx", "text/x-c++"),
    (".cs", "text/x-csharp"),
    (".php", "text/x-php"),
    (".pl", "text/x-perl"),
    (".pm", "text/x-perl"),
    (".r", "text/x-r"),
    (".R", "text/x-r"),
    (".lua", "text/x-lua"),
    (".dart", "text/x-dart"),
    (".ex", "text/x-elixir"),
    (".exs", "text/x-elixir"),
    (".erl", "text/x-erlang"),
    (".hrl", "text/x-erlang"),
    (".clj", "text/x-clojure"),
    (".cljs", "text/x-clojure"),
    (".hs", "text/x-haskell"),
    (".lhs", "text/x-haskell"),
    (".ml", "text/x-ocaml"),
    (".mli", "text/x-ocaml"),
    (".fs", "text/x-fsharp"),
    (".fsx", "text/x-fsharp"),
    // Shell/Scripts
    (".sh", "text/x-shellscript"),
    (".bash", "text/x-shellscript"),
    (".zsh", "text/x-shellscript"),
    (".fish", "text/x-shellscript"),
    (".ps1", "text/x-powershell"),
    (".psm1", "text/x-powershell"),
    (".bat", "text/x-batch"),
    (".cmd", "text/x-batch"),
    // Database
    (".sql", "text/x-sql"),
    (".pgsql", "text/x-sql"),
    (".mysql", "text/x-sql"),
    // Config files
    (".ini", "text/plain"),
    (".cfg", "text/plain"),
    (".conf", "text/plain"),
    (".config", "text/plain"),
    (".env", "text/plain"),
    (".env.local", "text/plain"),
    (".env.development", "text/plain"),
    (".env.production", "text/plain"),
    (".properties", "text/plain"),
    (".editorconfig", "text/plain"),
    (".gitignore", "text/plain"),
    (".gitattributes", "text/plain"),
    (".dockerignore", "text/plain"),
    (".npmignore", "text/plain"),
    (".eslintrc", "application/json"),
    (".prettierrc", "application/json"),
    (".babelrc", "application/json"),
    // Misc
    (".graphql", "text/x-graphql"),
    (".gql", "text/x-graphql"),
    (".proto", "text/x-protobuf"),
    (".tf", "text/x-terraform"),
    (".tfvars", "text/x-terraform"),
    (".dockerfile", "text/x-dockerfile"),
    (".makefile", "text/x-makefile"),
];

/// Special extension-less filenames → MIME type (v4 `specialFiles`).
const SPECIAL_FILES: &[(&str, &str)] = &[
    ("dockerfile", "text/x-dockerfile"),
    ("makefile", "text/x-makefile"),
    ("gemfile", "text/x-ruby"),
    ("rakefile", "text/x-ruby"),
    ("cmakelists.txt", "text/x-cmake"),
    ("license", "text/plain"),
    ("readme", "text/plain"),
    ("changelog", "text/plain"),
    ("authors", "text/plain"),
    ("contributors", "text/plain"),
];

/// Common text-based `application/*` MIME types (v4 `textApplicationTypes`).
const TEXT_APPLICATION_TYPES: &[&str] = &[
    "application/json",
    "application/xml",
    "application/javascript",
    "application/typescript",
    "application/x-javascript",
    "application/x-typescript",
    "application/x-sh",
    "application/x-shellscript",
    "application/graphql",
];

/// The standard-extension matcher `\.([^.]+)$` (v4's `.match(/\.([^.]+)$/)`).
static EXT_RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\.([^.]+)$").unwrap());

/// Whether a byte is printable ASCII or common whitespace (v4
/// `isPrintableOrWhitespace`).
///
/// Tab/newline/CR, printable ASCII (space..tilde), and every high byte
/// (`>= 0x80`) count as printable; nothing else does.
fn is_printable_or_whitespace(byte: u8) -> bool {
    if byte == 0x09 || byte == 0x0a || byte == 0x0d {
        return true;
    }
    if (0x20..=0x7e).contains(&byte) {
        return true;
    }
    if byte >= 0x80 {
        return true;
    }
    false
}

/// Detect if file content is likely plain text (v4 `isTextContent`).
///
/// Samples the first [`SAMPLE_SIZE`] bytes: an empty buffer is text; a 3rd null
/// byte short-circuits to binary; otherwise text iff the non-printable ratio is
/// strictly below [`NON_PRINTABLE_THRESHOLD`].
pub fn is_text_content(buffer: &[u8]) -> bool {
    let sample_length = buffer.len().min(SAMPLE_SIZE);

    if sample_length == 0 {
        // Empty files are considered text.
        return true;
    }

    let mut null_bytes: u32 = 0;
    let mut non_printable: u64 = 0;

    for &byte in &buffer[..sample_length] {
        // Null bytes are a strong indicator of binary content.
        if byte == 0x00 {
            null_bytes += 1;
            // More than a couple null bytes → likely binary.
            if null_bytes > 2 {
                return false;
            }
        }

        if !is_printable_or_whitespace(byte) {
            non_printable += 1;
        }
    }

    let non_printable_ratio = non_printable as f64 / sample_length as f64;
    non_printable_ratio < NON_PRINTABLE_THRESHOLD
}

/// Infer a MIME type from a filename (v4 `getMimeTypeFromExtension`).
///
/// Three passes over the lowercased name: (1) the first `ends_with` hit in
/// [`EXTENSION_MIME_MAP`] insertion order (handles `.d.ts`, `.env.local`);
/// (2) the `\.([^.]+)$` standard extension looked up exactly; (3) the special
/// extension-less basenames (`dockerfile`, `readme`, …). `None` if unknown.
pub fn get_mime_type_from_extension(filename: &str) -> Option<String> {
    let lower = filename.to_lowercase();

    // Check for compound extensions first (insertion order — first match wins).
    for &(ext, mime) in EXTENSION_MIME_MAP {
        if lower.ends_with(ext) {
            return Some(mime.to_string());
        }
    }

    // Check standard extension.
    if let Some(caps) = EXT_RE.captures(&lower) {
        let ext = format!(".{}", &caps[1]);
        return EXTENSION_MIME_MAP
            .iter()
            .find(|&&(k, _)| k == ext)
            .map(|&(_, v)| v.to_string());
    }

    // Handle special filenames without extensions.
    let base_name = lower.split('/').next_back().filter(|s| !s.is_empty());
    let base_name = base_name.unwrap_or(&lower);
    SPECIAL_FILES
        .iter()
        .find(|&&(k, _)| k == base_name)
        .map(|&(_, v)| v.to_string())
}

/// Whether a MIME type indicates text content (v4 `isTextMimeType`).
pub fn is_text_mime_type(mime_type: &str) -> bool {
    if mime_type.starts_with("text/") {
        return true;
    }
    TEXT_APPLICATION_TYPES.contains(&mime_type)
}

/// Detect if a file is text and infer its best MIME type (v4
/// `detectTextContent`).
pub fn detect_text_content(
    buffer: &[u8],
    filename: &str,
    provided_mime_type: &str,
) -> TextDetectionResult {
    let content_is_text = is_text_content(buffer);
    let extension_mime_type = get_mime_type_from_extension(filename);

    let is_generic = provided_mime_type == "application/octet-stream"
        || provided_mime_type.is_empty()
        || provided_mime_type == "application/x-unknown";

    let detected_mime_type: Option<String> = if extension_mime_type.is_some() && is_generic {
        extension_mime_type.clone()
    } else if content_is_text && is_generic {
        // Content is text but we don't recognize the extension.
        Some("text/plain".to_string())
    } else {
        None
    };

    let is_plain_text = content_is_text
        && (is_text_mime_type(provided_mime_type)
            || is_text_mime_type(detected_mime_type.as_deref().unwrap_or(""))
            || extension_mime_type.is_some());

    TextDetectionResult {
        is_plain_text,
        detected_mime_type,
        encoding: "utf-8".to_string(),
    }
}

/// Return the best MIME type for a file (v4 `getBestMimeType`).
///
/// The detected type when it is present AND non-empty (JS `||` truthiness),
/// otherwise the provided type.
pub fn get_best_mime_type(result: &TextDetectionResult, provided_mime_type: &str) -> String {
    match result.detected_mime_type.as_deref() {
        Some(m) if !m.is_empty() => m.to_string(),
        _ => provided_mime_type.to_string(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn empty_buffer_is_text() {
        assert!(is_text_content(&[]));
    }

    #[test]
    fn two_nulls_ok_three_binary() {
        // Two nulls in a 30-byte buffer: 2/30 < 0.10, so still text (the 3rd-null
        // short-circuit is not hit).
        let mut two = vec![b'a'; 30];
        two[2] = 0;
        two[17] = 0;
        assert!(is_text_content(&two));
        // Three nulls short-circuits to binary regardless of ratio.
        let three = b"\0\0\0aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa";
        assert!(!is_text_content(three));
    }

    #[test]
    fn exactly_ten_percent_nonprintable_is_not_text() {
        // 1 non-printable out of 10 = 0.10, which is NOT < 0.10 → not text.
        let mut buf = vec![b'a'; 9];
        buf.push(0x01);
        assert!(!is_text_content(&buf));
    }

    #[test]
    fn high_bytes_are_printable() {
        // High bytes are unconditionally printable (no UTF-8 validation).
        assert!(is_text_content(&[0xff, 0xfe, 0x80, b'a']));
    }

    #[test]
    fn ext_lookup_basic() {
        assert_eq!(
            get_mime_type_from_extension("foo.py").as_deref(),
            Some("text/x-python")
        );
        assert_eq!(get_mime_type_from_extension("foo.zzz"), None);
        assert_eq!(get_mime_type_from_extension("noext"), None);
    }

    #[test]
    fn ends_with_d_ts_resolves_via_ts() {
        // `.ts` is the first slice entry, so `ends_with` wins the compound pass.
        assert_eq!(
            get_mime_type_from_extension("component.d.ts").as_deref(),
            Some("text/typescript")
        );
    }

    #[test]
    fn special_file_and_path_basename() {
        assert_eq!(
            get_mime_type_from_extension("Dockerfile").as_deref(),
            Some("text/x-dockerfile")
        );
        assert_eq!(
            get_mime_type_from_extension("README").as_deref(),
            Some("text/plain")
        );
    }

    #[test]
    fn best_mime_type_falls_back_on_empty() {
        let r = TextDetectionResult {
            is_plain_text: false,
            detected_mime_type: Some(String::new()),
            encoding: "utf-8".to_string(),
        };
        assert_eq!(get_best_mime_type(&r, "provided/x"), "provided/x");
    }
}
